### ISSUE TITLE
Namespace for CMake library target is capitalized

### ISSUE THREAD

**Issue by Docteh** [lz4/lz4#1179]
Is there any particular reason the namespace is capitalized like `LZ4`?

Related to this PR: https://github.com/lz4/lz4/pull/1030 and specifically a recently added comment to the PR from Taichi https://github.com/lz4/lz4/pull/1030#discussion_r993075968

Since we noticed the change [recently](https://github.com/yuzu-emu/yuzu/pull/9054) I figure its better to make a new issue rather than comment on a PR from last year. 

In doing some distro sleuthing, I noticed that Gentoo ships the CMake files, and Arch and openSUSE tumbleweed doesn't. 

Edit: forgot the point.

Hoping for one of: `LZ4::lz4_shared` is right or ooops it should be `lz4::lz4_shared`

**Comment by eli-schwartz** [comment]
> In doing some distro sleuthing, I noticed that Gentoo ships the CMake files, and Arch and openSUSE tumbleweed doesn't.

This isn't surprising. The project officially supports GNU Makefiles and that's it, and exports a pkg-config file. So that's the only way to reliably detect lz4 availability.

External build systems including CMake are offered on an as-is basis, originally/primarily for IDE integration? and there are several of them, but only one of them provides CMake files. (Sadly, CMake makes it very hard to produce those files without using CMake to do it.) Ideally, you'd use something like Meson's `dependency()` interface that checks both pkg-config and `*-config.cmake` under the same name, and returns whichever one it can find. You can do that with a bit of if/else using CMake and FindPkgConfig.

**Comment by Cyan4973** [comment]
Let's assume we change now `LZ4::` to `lz4::`, to answer this issue.
What would be the consequences ?
Would existing projects relying on `LZ4::` break ?
Should we support both capitalization now ? Is that even possible ?

**Comment by dg0yt** [comment]
The capitalization inconsistency just became visible in updating the vcpkg port: 
https://github.com/microsoft/vcpkg/pull/40021
I found this issue when looking for existing discussions or decisions.

Notes:
- IMO inconsistent patterns are much harder to memorize.

- lz4 creates a package `lz4` (case-sensitive) with target `LZ4::lz4` (case-sensitive).

- vcpkg used to have a vendored `CMakeLists.txt` which created a package `lz4` (case-sensitive) with target `lz4::lz4`.
- The changes to downstream vcpkg ports indicate some use of `lz4::lz4`. I can't tell in general if this is from vendored find modules or driven by vcpkg.
  But I see that at least seastar has `Findlz4.cmake` i.e. package name `lz4` and target `lz4::lz4`.
- The pattern from official find modules is package name `Pkg` with target `Pkg::Pkg`. (But counterexamples do exist.)
- Some projects adopted config package patterns which allowed seamless switching from official find modules to config packages. (There is `CMAKE_FIND_PACKAGE_PREFER_CONFIG`.) This tends to support the pattern of consistent capitalization.
- VTK/Cmake/FindLZ4.cmake (also from Kitware) has package name `LZ4` and target `LZ4::LZ4` :roll_eyes: 

> Let's assume we change now `LZ4::` to `lz4::`, to answer this issue. 
> What would be the consequences ?
> Would existing projects relying on `LZ4::` break ?
> Should we support both capitalization now ? Is that even possible ?

Basically I don't want to promote a particular choice, but I think the current pattern should be replaced. Keeping the current package name of `lz4`, it might be easiest to go for `lz4::lz4`, implementing either this name or `LZ4::lz4` as an alias or interface target for the other one. In that case, 
- the config would support both capitalization variants,
- no CMake project should break. 
  (But I don't want to make claims about meson's cmake dependency lookup.)

**Comment by eli-schwartz** [comment]
meson's problem is a bit more subtle here. The canonical way to look up this software via the dependency provider installed by the Makefile build system -- the only "official" build system -- is to use:

```
# shell
pkg-config --cflags --libs liblz4

# autotools
PKG_CHECK_MODULES([LZ4], [liblz4])

# cmake
find_package(PkgConfig)
pkg_check_modules(LZ4 REQUIRED IMPORTED_TARGET GLOBAL liblz4)
target_link_libraries(foobar PkgConfig::LZ4)

# meson
lz4_dep = dependency('liblz4')
```

Checking for the cmake files as well, you can do this:

```
# cmake
find_package(lz4)
if(TARGET LZ4::lz4_shared)
    target_link_libraries(foobar LZ4::lz4_shared)
else()
    find_package(PkgConfig)
    pkg_check_modules(LZ4 REQUIRED IMPORTED_TARGET GLOBAL liblz4)
    target_link_libraries(foobar PkgConfig::LZ4)
endif()

# meson
lz4_dep = dependency('liblz4', 'lz4')
```

Meson will check for liblz4 with pkg-config, and with cmake. It will always fail to detect `find_package(liblz4)`, and if pkg-config isn't installed (as is common on Windows, where it isn't shipped by the base system), meson will try the second name.

So it will then check for lz4 with pkg-config, and always fail because that name never exists. Then it checks for `find_package(lz4)`, and very likely does succeed.

Meson then needs to figure out what the name of `LZ4::lz4_shared` is. Currently, if you try it, it prints this:

```
WARNING: Could not find and exact match for the CMake dependency lz4.

However, Meson found the following partial matches:

    ['LZ4::lz4_shared']

Using imported is recommended, since this approach is less error prone
and better supported by Meson. Consider explicitly specifying one of
these in the dependency call with:

    dependency('lz4', modules: ['lz4::<name>', ...])

Meson will now continue to use the old-style lz4_LIBRARIES CMake
variables to extract the dependency information since no explicit
target is currently specified.


Run-time dependency lz4 found: NO 

meson.build:31:0: ERROR: Dependency lookup for lz4 with method 'cmake' failed: Dependency lz4 not found: CMake: failed to guess a CMake target for lz4.
Try to explicitly specify one or more targets with the "modules" property.
Valid targets are:
['LZ4::lz4_shared']
```

This is because it doesn't define a target which, when converted to lowercase, matches `lz4::lz4`. It *also* doesn't define the old-style lz4_LIBRARIES variable, so the fallback fails too.

As a result, all meson users relying on the existing cmake support will have to specify:

```
lz4_dep = dependency('liblz4', 'lz4', method: 'cmake', modules: ['lz4::lz4_shared'])
```
and pray it's not built statically. As long as that target continues to exist, even if other targets with different capitalization are added, existing use cases will still work.

So:

> Basically I don't want to promote a particular choice, but I think the current pattern should be replaced. Keeping the current package name of `lz4`, it might be easiest to go for `lz4::lz4`, implementing either this name or `LZ4::lz4` as an alias or interface target for the other one. In that case,

This will be purely a win for meson as well. But I would like to recommend additionally adding liblz4Config.cmake, possibly *also* while preserving the existing lz4Config.cmake as an alias, to simplify things even further: the name "liblz4" would always work first try, no need to probe multiple names.

**Comment by Cyan4973** [comment]
Something I don't know:
is it possible to support both `LZ4::` and `lz4::` ?

This would ensure that systems already looking for `LZ4::` don't break.

**Comment by eli-schwartz** [comment]
It is certainly possible if you use a handwritten lz4Targets.cmake rather than the cmake code `install(EXPORT lz4Targets ....)`.

I will defer to someone that knows a lot more about cmake than I do, to answer the question of whether it is possible to do it via cmake's builtin generator.

**Comment by dg0yt** [comment]
> Is it possible to support both `LZ4::` and `lz4::` ?
> 
> This would ensure that systems already looking for `LZ4::` don't break.

Actually it is trivial. And maybe we could use a small window of opportunity to test it end-to-end in vpckg.
